docs/tracker-query.1
.TH tracker-query 1 "July 2007" GNU "User Commands"

.SH NAME
tracker-query \- command line tool to query tracker database 

.SH SYNOPSIS
.B tracker-query
.I [OPTIONS] query-file.rdf [MetaDataFields...]

.SH DESCRIPTION
.B tracker-query
Exectutes an RDF query and prints the result. Additionally the behavior
can be customized using below described options.

.SH OPTIONS
.TP
\-?, --help
Show a brief help message.

.TP
\-s, --service=ServiceName
Search for files from a specific service. ServiceName has to be one of
trackers services.  For mor information on services, see
.BR tracker-services (7).

.TP
\-t, --search-term=SerchTerm
Adds a full text search filter to the query.

.TP
\-k, --keyword=Keyword
Adds a keyword filter to the query.

.SH "SEE ALSO"
.BR trackerd (1),
.BR tracker-services (7).


NEWS
Tracker 0.6.6
============= (03 Mar 2008)

  Applet
  * HIG fix for preference window.
  * Progress for index merges added.

  Tracker daemon
  * Fixed shutdown of daemon when paused to prevent hangs.
  * Fixed an unfinished final index merge step when next restarted.



Tracker 0.6.5
============= (27 Feb 2008)

  Applet
  * Removed popup window and replaced with tooltip displaying status
    and progress.
  * New Applet preferences to turn off animation/hide icon/set
    smart pause.
  * Smart pause feature which can auto-pause the Tracker daemon
    whenever the user presses a key or moves the mouse, thereby
    improving system responsiveness during indexing.
  * Annoying merge messages have now been removed.
  * Reindex option added.
  * About box added.
  * Now hidden if indexing is disabled.
  * Bugs with constant animation fixed.
  * Animation rate slowed right down so as to be less annoying.

  Tracker daemon
  * Temporary blacklisting of frequently changing files (they are
    indexed after one hour or on next restart of Tracker daemon).
    This prevents constant indexing from downloads and torrents.
  * More blacklisting of common tmp formats and vmware files.
  * Memory leak fixes.
  * Removed CPU-heavy corruption scan which can last long time and
    occurs on battery too at startup when tracker daemon was not
    previously shut down properly. Instead we now use bigger
    transactions with fsync on to prevent corruption in the first
    place (only slightly more disk IO is used during indexing as a
    result).
  * Fixed crashers.
  * Fixed D-Bus shutdown/reindex call.



Tracker 0.6.4
============= (11 Dec 2007)

  * Made indexing more robust by pausing if disk space is low or
    index grows too big.
  * Limit log file size to 10MB.
  * New Tracker Applet - animates when indexing, provides ability to
    pause indexing as well as viewing status and progress feedback
    from indexer, statistics, and notification warnings; it can
    also launch Tracker Preferences and Tracker Search Tool.
  * New power management options enable much better customization.
  * Ignored files fixes.
  * Deskbar/tracker integration fixes.
  * Made most prefs live and affect Tracker in real time; others
    will prompt for restart and/or reindex where necessary.
  * Shell script fixes
  * Fixed Imap bug with embedded Auth in URI.
  * Built-in corruption check and scan when Tracker is not shut down
    cleanly - prevents infinite looping.
  * Fix index deletions.
  * Many more bug fixes and stability improvements.



Tracker 0.6.3
============= (25 Sep 2007)

  * Fixed memory leaks.
  * New index merging which allows for much better performance and
    much less disk hogging I/O.
  * Battery throttling - Tracker now stops indexing when on battery.
  * Many more optimisations - transactions, use of O_NOATIME and
    posix_fadvise everywhere.
  * Check files exist before including in search results.
  * Fixed disabling of indexing option in tracker-preferences.
  * HIG fixes to tracker-preferences screen.
  * Fixed resetting of stats.

  note:
    This version will cause your hard drive to be re-indexed so
    that stats are fixed (previous version broke them).



Tracker 0.6.2
============= (25 Sep 2007)

  * New Sqlite-based indexer which utilises the new incremental blob
    I/O in sqlite 3.4.
  * Highly-optimised email indexing (up to 5x faster).
  * Dramatically reduced disk access and disk contention.
  * Indexer now pauses for a grace period when non-Tracker processes
    write to disk (providing changed files are being watched by
    Tracker) - minimises slowdowns when compiling or checking out
    source code.
  * Makes use of idle class disk IO scheduling if available.
  * Makes preliminary use of NO_ATIME (some disk access still uses
    fopen).
  * Fixed restore of user metadata on re-index (keywords are
    auto-restored).
  * Added increased number of (junk) files to automatically ignore.
  * Improved stopwords.
  * New Deskbar handler that uses the new Deskbar api (2.19+).
  * Fixed old Deskbar handler to remove race condition causing
    crashes.
  * Fixed a number of annoying bugs in email indexing and Tracker
    Preferences.

  note:
    This version will cause your hard drive to be re-indexed due to
    the new Sqlite indexer backend.


Tracker 0.6.1
============= (08 Aug 2007)

  * translations and bugfixes


Tracker 0.6.0 "better late than never"
============= (24 Jul 2007)

  * Evolution Emails (mbox/pop/imap) now indexed.
  * Gaim/Pidgin chat logs now indexed.
  * Applications now indexed.
  * New files indexed - all spreadsheets, csv (these require
    Gnumeric to be installed) and oo draw.
  * Vastly improved and new search UI using a sidebar for categories
    and a new metadata tile.
  * Tagging support in the UI.
  * New tracker-preferences UI for easy setting of prefs.
  * Added libtracker-gtk widget toolkit.
  * Now uses XDG directories.
  * Optimised triple store.
  * Handles multiple triple stores (files, emails and user data are
    all separate dbs).
  * New IDF based ranking algorithm.
  * XMP metadata support added.
  * Made use of async calls in tracker-search-tool to prevent it
    from freezing.
  * Added a 60 second delay before indexing to prevent slowdowns on
    desktop start-up.
  * Added support for crawling directories without watching overhead
    (useful for indexing source code without causing slowdowns when
     compiling).


Tracker 0.5.4 "Speed Daemon"
============= (26 Jan 2007)

  * Indexing at ludicrous speed - massively optimised indexing so
    its now 10x faster than previous version. Indexing speeds are
    now around 100 text files per second (which is about the
    maximum possible considering the I/O time to read 100 files from
    a hard drive).
  * Index while you work - now scheduled even better so it should
    never slow the system down whilst allowing users to work without
    being affected by or even noticing it.

src/libtracker/tracker-query.c
/* Tracker - indexer and metadata database engine
 * Copyright (C) 2006, Mr Jamie McCracken (jamiemcc@gnome.org)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA  02110-1301, USA.
 */

#include <locale.h>
#include <sys/param.h>
#include <stdlib.h>
#include <glib.h>
#include <glib-object.h>
#include <glib/gstdio.h>

#include "../libtracker/tracker.h" 

#include <config.h>
#ifdef OS_WIN32
#include "../trackerd/mingw-compat.h"
#endif

static gchar *search = NULL;
static gchar **fields = NULL;
static gchar *service = NULL;
static gchar *keyword = NULL;

static GOptionEntry entries[] = {
	{"service", 's', 0, G_OPTION_ARG_STRING, &service, "search from a specific service", "service"},
	{"search-term", 't', 0, G_OPTION_ARG_STRING, &search, "adds a fulltext search filter", "search-term"},
	{"keyword", 'k', 0, G_OPTION_ARG_STRING, &keyword, "adds a keyword filter", "keyword"},
	{G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_STRING_ARRAY, &fields, "Metadata Fields", NULL},
	{NULL}
};


static int field_count;

static char *
realpath_in_utf8 (const char *path)
{
	char *str_path_tmp = NULL, *str_path = NULL;

	str_path_tmp = realpath (path, NULL);

	if (str_path_tmp) {
		str_path = g_filename_to_utf8 (str_path_tmp, -1, NULL, NULL, NULL);

		g_free (str_path_tmp);
	}

	if (!str_path) {
		g_warning ("realpath_in_utf8(): locale to UTF8 failed!");
		return NULL;
	}

	return str_path;
}

static void
get_meta_table_data (gpointer value)
		    
{
	char **meta, **meta_p;

	meta = (char **)value;

	int i = 0;
	for (meta_p = meta; *meta_p; meta_p++) {

		char *str;

		str = g_filename_from_utf8 (*meta_p, -1, NULL, NULL, NULL);

		if (i == 0) {
			g_print ("%s : ", str);

		} else {
			g_print ("%s, ", *meta_p);
		}
		i++;
	}
	g_print ("\n");
}



int
main (int argc, char **argv) 
{
	GOptionContext *context = NULL;
	ServiceType type;
	char **p_strarray;

	char *buffer, *tmp;
	gsize buffer_length;
	GPtrArray *out_array = NULL;
	GError *error = NULL;
	TrackerClient *client = NULL;


	setlocale (LC_ALL, "");

	context = g_option_context_new ("RDFQueryFile [MetaDataField...] ... - perform an rdf query and return results witrh specified metadata fields");
	g_option_context_add_main_entries (context, entries, NULL);
	g_option_context_parse (context, &argc, &argv, &error);

	if (error) {
		g_printerr ("invalid arguments: %s\n", error->message);
		return 1;
	}

	if (!fields) {
		g_printerr ("missing input rdf query file, try --help for help\n");
		return 1;
	}
	

	if (!service) {
		type = SERVICE_FILES;
	} else {
		type = tracker_service_name_to_type (service);

		if (type == SERVICE_OTHER_FILES && g_ascii_strcasecmp (service, "Other")) {
			g_printerr ("service not recognized, searching in Other Files...\n");
		}
	}
	
	char *str_path = realpath_in_utf8 (fields[0]);

	if (!str_path) {
		return 1;
	}

	if (!g_file_get_contents (str_path, &tmp, &buffer_length, NULL)) {
		g_print ("Could not read file %s\n", str_path);
		return 1;
	}

	g_free (str_path);

	buffer = g_locale_to_utf8 (tmp, buffer_length, NULL, NULL, NULL);

	if (!buffer) {
		g_warning ("Cannot convert query file to UTF8!");
		g_free (tmp);
		return 1;
	}


	client =  tracker_connect (FALSE);

	if (!client) {
		g_print ("Could not initialise Tracker over dbus connection - exiting...\n");
		return 1; 
	}

	char **meta_fields = NULL;

	g_free (fields[0]);
	int i = 0;
	for (p_strarray = fields+1; *p_strarray; p_strarray++) {
		fields[i] = *p_strarray;
		i++;
	}
	fields[i] = NULL;


	if (i == 0) {
		meta_fields = g_new (char *, 2);

		field_count = 1;

		meta_fields[0] = g_strdup ("File:Mime");
		meta_fields[1] = NULL;
		g_strfreev  (fields);
		fields = meta_fields;
	} 

	out_array = tracker_search_query (client, -1, type, fields, search, keyword, buffer, 0, 512, FALSE, &error);
	
	g_strfreev (meta_fields);

	if (error) {
		g_warning ("An error has occurred : %s\n", error->message);
		g_error_free (error);
		g_free (buffer);
		return 1;
	}

	if (out_array) {
		g_ptr_array_foreach (out_array, (GFunc)get_meta_table_data, NULL);
		g_ptr_array_free (out_array, TRUE);
	}


	tracker_disconnect (client);	

	g_free (buffer);

	return 0;
}
